80 Update widgets to use MUI base

abigailalexander opened the pull request:
This PR is the culmination of several smaller PRs to update the widgets listed in #80 to use MUI. Other points to note:

- A MUI theme has been created that has some Phoebus defaults for colours and font. Client applications can override this by passing a theme object into an EmbeddedDisplay. Styling set in the .bob file will also override the default theming
- Tests have been updated as MUI doesn't expose styles in the same way as normal HTML elements. This means snapshots no longer accurately reflect styling, so we rely on normal unit tests to check style
- Rotation property has been added to relevant widgets such as Label, Readback and ActionButton. It will not work if the height and width passed in are relative (aka `height = 5vh`) as the calculation doesn't work.
- The input component (different to the Input widget) has been moved to cs-web-proto, as this is the only place it's used. Given that it isn't a CSStudio widget, I think it's better to not include it in this library

abigailalexander commented on src/ui/widgets/Input/input.tsx:
```diff
@@ -1,102 +1,195 @@
+      inset: "0px"
+    },
+    "&.Mui-disabled": {
+      cursor: "not-allowed",
```
Oddly, I notice this behaviour in Phoebus as well which would suggest it might be intentional? The cs-web-lib widget only shows the cursor on left and right side borders, whereas Phoebus only shows it on all of the borders.

rjwills28 replied:
I think this is actually a bug in Phoebus as this occurs only when the 'Multi-Line' property has been enabled for this widget. If you switch that property to disabled you should see the 'not-allowed' cursor over the whole widget? I'll look into this on the Phoebus end.

abigailalexander commented on src/ui/widgets/ActionButton/actionButton.tsx:
```diff
@@ -11,48 +11,102 @@ import {
+    wordBreak: "break-word",
+    textTransform: "none"
+  },
+  "&.Mui-disabled": {
```
As mentioned, I looked into this but MUI makes this quite difficult to overwrite. It applies a transparent black across the entire widget using `-webkit-text-fill-color`, which is what causes the greying out. Since it's overlaying another colour and not modifying the one we've set, I struggled to get it to do what I wanted. I've left this as is for now.

abigailalexander commented on src/ui/widgets/Input/input.tsx:
```diff
@@ -1,102 +1,195 @@
+    alignmentV = "end";
   }
 
+  const backgroundColor = transparent
```
It seems like we have been passing in the alpha value from Phoebus .bob files but it wasn't being applied because alpha in HTML/CSS must be between 0-1. I've added a step in parsing colour to transform it from 0-255 to 0-1, and default to 1 if none is passed in. This allows us to make use of transparency here now.

abigailalexander commented on src/ui/widgets/MenuButton/menuButton.tsx:
```diff
@@ -21,129 +22,165 @@ import {
 
   // Store whether component is disabled or not
-  let disabled = readonly;
+  let disabled = !enabled;
```
I've updated this now to use the proper MUI disabled functionality.

abigailalexander commented on src/ui/widgets/Readback/readback.tsx:
```diff
@@ -34,41 +34,72 @@ const ReadbackProps = {
   backgroundColor: ColorPropOpt,
   border: BorderPropOpt,
-  rotationAngle: FloatPropOpt,
+  rotationStep: FloatPropOpt,
```
I couldn't seem to recreate this as I had no issues rotating text? It's possible this was caused by the same issue as for the actionbutton, where we set dimensions to "100%" for default widget size. This has been corrected now.

abigailalexander commented:
Thanks a lot for the review! I think I've fixed all of the issues mentioned now, except for the disabled font. Regarding the Slider scale, I've now added one in but MUI does not include it by default so we have to manually calculate it. It isn't very obvious how Phoebus calculates the tick sizes so there may be some differences between the exact scale depending on tick step size etc.

rjwills28 commented on src/ui/widgets/SlideControl/slideControl.tsx:
```diff
@@ -1,115 +1,176 @@
-  if (!editing && inputValue !== stringValue) {
-    setInputValue(stringValue);
-  }
+  const marks = showScale
```
This is looking good - thanks for the work on getting the scale to show. I don't think we need to worry that the tick interval size might be slightly different to Phoebus. I think Phoebus is probably different to CS-Studio anyway.

One minor issue is that if the property 'show scale' is false then you also lose the alarm indicators, e.g. HIHI, LOLO etc. I think this is because they are added to the same `marks` variable, which is only displayed if 'show scale' is true. Would it be possible to disentangle these so that you can not show a scale but still show the alarms?

rjwills28 reviewed (changes requested):
Thanks so much for making all these changes and fixing these issues. I've resolved all bar one existing issue relating to the slider widget and just added one more that I found relating to the ActionButton.

abigailalexander commented:
Thanks for looking over it again, I have hopefully fixed these issues now (and also another minor issue with the arc widget that I spotted while testing).

rjwills28 reviewed (approved):
Thanks for addressing all the issues mentioned! Behaviour looks good!
